<h1 dir="rtl" style="font-family: Vazir; width: 85%;">تمرین شماره ۳
: درس بازیابی هوشمند اطلاعات - دانشگاه تهران، پائیز ۱۴۰۳</h1>

<div dir="rtl" style="font-family: Vazir; width: 85%; font-size: 18px;">
نام: عرفان شهابی
<br/>


<div dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
سوالات خودتان را می‌توانید از طریق ایمیل
<code>adibialii@ut.ac.ir</code>
 از طراح تمرین ۳ بپرسید.

<div dir="rtl" style="font-family: Vazir; width: 85%; font-size: 18px; color: red; font-weight: bold;">
قوانین و توضیحاتی آخر فایل تمرین حتما به دقت مطالعه شود.
</div>

# **سوال اول:**

In [2]:
#YOUR CODE HERE
import kagglehub

# Download latest version
path = kagglehub.dataset_download("lakshmi25npathi/imdb-dataset-of-50k-movie-reviews")

print("Path to dataset files:", path)

100%|██████████| 25.7M/25.7M [01:57<00:00, 229kB/s] 

Extracting files...
Path to dataset files: /Users/erfanshahabi/.cache/kagglehub/datasets/lakshmi25npathi/imdb-dataset-of-50k-movie-reviews/versions/1


الف

<div style="direction: rtl; text-align: right;">
در این کد ابتدا کتابخانه های مورد نیاز ایمپورت شده اند. کتابخانه pandas جهت پردازش داده های جدولی، کتابخانه nlkt برای توکن سازی، حذف استاپ ورد ها و یافتن stem کلمات. از کتابخانه beautifulsoup هم برای حذف تگ  های html استفاده کردم.
ابتدا لیست استاپ ورد های انگلیسی و مدل های مورد نیاز برای توکنایز کردن متن دانلود شده است.
سپس دیتاست نظرات کاربران را از مسیر مشخص شده آپلود کرده ام. در مرحله بعد لیست استاپ وردها را به صورت set ذخیره کردم. سپس از porterstemmer برای پیدا کردن stem کلمات را جدا کردم.
در مرحله بعد با استفاده از preprocess_text ابتدا تگ های html را در متن حذف کردم سپس متن ورودی را به کوچکترین شکل ممکن توکنایز کردم و کلمات غیر عددی بدون استاپ ورد را استخراج کردم.
در مرحله بعد نظرات مثبت و منفی با استفاده از فیلتر کردن بر اساس ستون sentiment استخراج می‌شوند و در دو DataFrame جداگانه ذخیره می‌شوند (positive_reviews و negative_reviews).
در پایان نظرات مثبت چاپ میشوند.
نظرات منفی هم در رشته کد بعدی چاپ میشوند.
تعداد نظرات مثبت و منفی برابر ۱۲۵۰۰ است.

In [34]:
import pandas as pd
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import PorterStemmer
from bs4 import BeautifulSoup  

nltk.download('stopwords')
nltk.download('punkt')


data_path = '/Users/erfanshahabi/.cache/kagglehub/datasets/lakshmi25npathi/imdb-dataset-of-50k-movie-reviews/versions/1/IMDB Dataset.csv'


df = pd.read_csv(data_path)


stop_words = set(stopwords.words('english'))
stemmer = PorterStemmer()


def clean_html(text):
    return BeautifulSoup(text, "html.parser").get_text()


def preprocess_text(text):
   
    text = clean_html(text)
   
    tokens = word_tokenize(text.lower())
    filtered_tokens = [stemmer.stem(word) for word in tokens if word.isalnum() and word not in stop_words]
    return filtered_tokens


df['processed_text'] = df['review'].apply(preprocess_text)


positive_reviews = df[df['sentiment'] == 'positive']
negative_reviews = df[df['sentiment'] == 'negative']


print(positive_reviews)


[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/erfanshahabi/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     /Users/erfanshahabi/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
/var/folders/mv/7bx_cn1n6zb77rddr70qj7fm0000gn/T/ipykernel_2969/11635765.py:23: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.
  return BeautifulSoup(text, "html.parser").get_text()


                                                  review sentiment  \
0      One of the other reviewers has mentioned that ...  positive   
1      A wonderful little production. <br /><br />The...  positive   
2      I thought this was a wonderful way to spend ti...  positive   
4      Petter Mattei's "Love in the Time of Money" is...  positive   
5      Probably my all-time favorite movie, a story o...  positive   
...                                                  ...       ...   
49983  I loved it, having been a fan of the original ...  positive   
49985  Imaginary Heroes is clearly the best film of t...  positive   
49989  I got this one a few weeks ago and love it! It...  positive   
49992  John Garfield plays a Marine who is blinded by...  positive   
49995  I thought this movie did a down right good job...  positive   

                                          processed_text  
0      [one, review, mention, watch, 1, oz, episod, h...  
1      [wonder, littl, product, film, tec

In [36]:
print(negative_reviews)

                                                  review sentiment  \
3      Basically there's a family where a little boy ...  negative   
7      This show was an amazing, fresh & innovative i...  negative   
8      Encouraged by the positive comments about this...  negative   
10     Phil the Alien is one of those quirky films wh...  negative   
11     I saw this movie when I was about 12 when it c...  negative   
...                                                  ...       ...   
49994  This is your typical junk comedy.<br /><br />T...  negative   
49996  Bad plot, bad dialogue, bad acting, idiotic di...  negative   
49997  I am a Catholic taught in parochial elementary...  negative   
49998  I'm going to have to disagree with the previou...  negative   
49999  No one expects the Star Trek movies to be high...  negative   

                                          processed_text  
3      [basic, famili, littl, boy, jake, think, zombi...  
7      [show, amaz, fresh, innov, idea, 7

ب

<div style="direction: rtl; text-align: right;">
در این کد سعی شده ۵۰ کلمه پرکاربرد در نظرات استخراج شوند. 
ابتدا counter از کتابخانع collections ایمپورت شده است که از آن برای شمارش فراوانی کلمات استفاده می شود.
سپس تابع get_top_words تعریف شده است که نظرات و تعداد کلمات پرکاربرد درخواستی را به عنوان ورودی می گیرد.
ابتدا تمامی کلمات موجود در نظرات جمع‌آوری می‌شوند. این کار با استفاده از یک لیست فشرده انجام می‌شود که برای هر نظر، تمام کلمات آن را استخراج می‌کند.
سپس از Counter برای شمارش فراوانی هر کلمه استفاده شده و n کلمه پرکاربردتر را برمی‌گرداند.
در مرحله بعد با فراخوانی تابع get_top_words برای نظرات مثبت (positive_reviews['processed_text']) و نظرات منفی (negative_reviews['processed_text'])، ۵۰ کلمه پرکاربرد از هر دسته استخراج می‌شود.
برای ذخیره کلمات پرکاربرد در فایل‌ها، دو فایل متنی به نام‌های top_words_positive.txt و top_words_negative.txt ایجاد می‌شود و کلمات در آن ها ذخیره میشوند که در فایل ارسالی در elearn قابل مشاهده هستند.

In [37]:
from collections import Counter

def get_top_words(reviews, n=50):
    all_words = [word for review in reviews for word in review]
    return Counter(all_words).most_common(n)


top_positive_words = get_top_words(positive_reviews['processed_text'])
top_negative_words = get_top_words(negative_reviews['processed_text'])


with open('top_words_positive.txt', 'w') as pos_file, open('top_words_negative.txt', 'w') as neg_file:
    print("Top 50 Words in Positive Reviews:")
    for word, count in top_positive_words:
        print(f"{word}: {count}")
        pos_file.write(f"{word}: {count}\n")
    
    print("\nTop 50 Words in Negative Reviews:")
    for word, count in top_negative_words:
        print(f"{word}: {count}")
        neg_file.write(f"{word}: {count}\n")


Top 50 Words in Positive Reviews:
film: 49345
movi: 43477
one: 27142
like: 19972
time: 15600
see: 14738
good: 14676
stori: 13725
charact: 13704
make: 13176
great: 12887
get: 12578
watch: 12568
love: 12114
well: 11435
would: 11243
show: 10865
realli: 10694
also: 10590
play: 9894
scene: 9765
even: 9761
much: 9034
end: 8921
first: 8897
peopl: 8692
way: 8579
think: 8409
best: 8340
go: 8071
look: 7904
mani: 7562
life: 7534
year: 7307
work: 7290
could: 7168
know: 7152
made: 7037
thing: 6907
two: 6901
perform: 6887
take: 6713
seen: 6638
act: 6624
say: 6616
come: 6572
littl: 6392
man: 6376
still: 6343
never: 6270

Top 50 Words in Negative Reviews:
movi: 56338
film: 43359
one: 25879
like: 24048
make: 15422
even: 15225
would: 15211
get: 15159
watch: 14651
time: 14490
good: 14429
bad: 14233
charact: 14035
see: 12947
realli: 12284
look: 11350
could: 11333
scene: 11086
stori: 10661
act: 10222
much: 9936
go: 9648
end: 9432
peopl: 9400
thing: 9321
think: 8987
show: 8639
made: 8415
say: 8270
plot: 824

ج

<div style="direction: rtl; text-align: right;">
در این کد تلاش کردم ۵۰ جفت کلمه پر تکرار در نظرات منفی و مثبت را استخراج کنم.
ابتدا تابع get_top_bigrams رو تعریف کردم که این تابع به ازای هر نظر در دیتاست تمام بای گرم های موجود در نظر را استخراج میکند. برای این کار از list comprehension استفاده شده است که تمامی بای گرم‌ ها را برای هر نظر استخراج می‌کند سپس از کانتر برای شمارش فراوانی جفت کلمات استفاده می کند. در نهایت با استفاده از most_common(50)، ۵۰ جفت کلمه پر کاربرد را استخراج می کند.
سپس با استفاده از این تابع ۵۰ جفت کلمه پر کاربرد در نظرات مثبت و منفی را محاسبه کرده و در فایل مربوطه ذخیره میکند و ان ها را چاپ هم میکند.

In [38]:
def get_top_bigrams(reviews, n=50):
    all_bigrams = [tuple(review[i:i+2]) for review in reviews for i in range(len(review)-1)]
    return Counter(all_bigrams).most_common(n)


top_positive_bigrams = get_top_bigrams(positive_reviews['processed_text'])
top_negative_bigrams = get_top_bigrams(negative_reviews['processed_text'])

with open('top_bigrams_positive.txt', 'w') as pos_file, open('top_bigrams_negative.txt', 'w') as neg_file:
    print("\nTop 50 Bigrams in Positive Reviews:")
    for bigram, count in top_positive_bigrams:
        print(f"{bigram}: {count}")
        pos_file.write(f"{bigram}: {count}\n")
    
    print("\nTop 50 Bigrams in Negative Reviews:")
    for bigram, count in top_negative_bigrams:
        print(f"{bigram}: {count}")
        neg_file.write(f"{bigram}: {count}\n")



Top 50 Bigrams in Positive Reviews:
('one', 'best'): 1633
('watch', 'movi'): 1273
('even', 'though'): 1054
('see', 'movi'): 945
('ever', 'seen'): 943
('look', 'like'): 927
('first', 'time'): 909
('new', 'york'): 832
('main', 'charact'): 825
('highli', 'recommend'): 816
('movi', 'like'): 807
('special', 'effect'): 799
('watch', 'film'): 773
('year', 'old'): 763
('great', 'movi'): 752
('year', 'ago'): 749
('film', 'like'): 740
('good', 'movi'): 734
('see', 'film'): 719
('like', 'movi'): 701
('love', 'movi'): 680
('take', 'place'): 641
('movi', 'ever'): 639
('make', 'film'): 639
('make', 'movi'): 626
('great', 'film'): 621
('horror', 'film'): 613
('well', 'done'): 605
('love', 'stori'): 592
('movi', 'one'): 589
('want', 'see'): 583
('real', 'life'): 576
('film', 'one'): 575
('year', 'later'): 573
('pretti', 'good'): 567
('saw', 'movi'): 567
('must', 'see'): 554
('would', 'like'): 552
('high', 'school'): 536
('worth', 'watch'): 534
('realli', 'good'): 533
('fall', 'love'): 521
('one', 'fa

د

<div style="direction: rtl; text-align: right;">
در قسمت ب پر استفاده ترین کلمات در نظرات مثبت کلماتی مانند film, movie, one, like, time, see,... هستند که منطقی هم به نظر می رسند. جملاتی مانند "it was one of the best movies that i havr seen", "i liked this film", ... از این کلمات استفاده میکنند و استفاده از قید ها میتواند نشان دهنده مثبت بودن نظر کاربر نسبت به فیلم باشد.
از طرفی پر استفاده ترین کلمات در نظرات منفی نیز شامل کلمات movie, film, one, like, make, even, would, .... هستند که میتوانند از جملاتی مانند "it was the worst movie that i have seen", "i would never see this film again", "this filme makes me angry",... میتواند از نمونه های این نظرات باشند. وجود استم های film, movi میتواند به این دلیل باشد که کاربران چه با نظر مثبت و چه با نظر منفی چون درباره فیلم صحبت میکنند از این وااژه ها زیاد استفاده میکندد.
<br><br>
در قسمت ج جفت کلمات نظرات مثبت مانند 
('one', 'best'): 1633
('watch', 'movi'): 1273
('even', 'though'): 1054
('see', 'movi'): 945
('ever', 'seen'): 943
('look', 'like'): 927
('first', 'time'): 909 
بیشترین تکرار را با هم دارند که بسیار قابل انتظار است. one best میتواند نماینده نظر one of the best movies باشد.
و در این قسمت جفت کلمات نظرات منفی کلماتی شامل:
('look', 'like'): 2758
('watch', 'movi'): 1755
('ever', 'seen'): 1643
('wast', 'time'): 1617
('special', 'effect'): 1434
('movi', 'ever'): 1320
هستند که برخی از آن ها کاملا قابل توجیه هستند مثلا movi ever, ever seen میتواند نماینده جمله it was the worse movie that i ever seen


# **سوال دوم:**

In [10]:
#YOUR CODE HERE
import kagglehub

# Download latest version
path = kagglehub.dataset_download("snap/amazon-fine-food-reviews")

print("Path to dataset files:", path)

100%|██████████| 242M/242M [12:38<00:00, 335kB/s]  

Extracting files...


Path to dataset files: /Users/erfanshahabi/.cache/kagglehub/datasets/snap/amazon-fine-food-reviews/versions/2


الف

<div style="direction: rtl; text-align: right;">
در این کد سعی شده گراف رابطه بین محصولات و کاربران رسم شود که کاربران و محصولات نود های آن هستند و یال های آن با استفاده از نظراتی که کاربر برای آن محصول ثبت کرده است رسم میشود.
ابتدا کتابخانه های پانداس و networkx ایمپورت شده اندو کتابخانه networkx برای ساخت و مدیریت استفاده می شود.
سپس فایل دیتا ست آدرس دهی شده و فایل خوانده می شود و ستون های مربوط به کاربران، محصپلات و امتیاز کاربر به محصول جدا شدند. در مرحله بعد یک گراف جهت دار ایجاد میشود که در این گراف کاربران و محصولات به عنوان نود های جداگانه خواهند بود.
سپس براای هر ردیف داده ها که نمایانگر یک نظر هستند یک یال بین کاربر و محصول کشیده می شود که وزن آن یال score است.
گزاف رسم شده باید گرافی جهت دار باشد چرا که یال ها از کاربر به محصول کشیده می شود همچنین این گراف وزن دار است و وزن یال ها برابر score میباشند و این گراف شامل دو نوع نود کاربر و محصول است.

In [20]:
import pandas as pd
import networkx as nx


data_path = '/Users/erfanshahabi/.cache/kagglehub/datasets/snap/amazon-fine-food-reviews/versions/2/Reviews.csv'
df = pd.read_csv(data_path)

df = df[['UserId', 'ProductId', 'Score']]


G = nx.DiGraph()


for _, row in df.iterrows():
    user = row['UserId']
    product = row['ProductId']
    score = row['Score']
    G.add_edge(user, product, weight=score)


print(f"Graph created with {G.number_of_nodes()} nodes and {G.number_of_edges()} edges.")


Graph created with 330317 nodes and 560804 edges.


ب و ج

<div style="direction: rtl; text-align: right;">
در این کد با استفاده از کتابخانه networkx الگوریتم hits ۱۰ بار تکرار شده و مقادیر hub و authorities برای هر نود محاسبه میشود.
سپس در رشته کد بعدی تعداد یال هایی که به هر نود وارد و یا از هر نود خارج میش.وند را محاسبه کردم. این مقادیر به صورت دیکشنری تبدیل شده‌اند که در آن هر گره به عنوان کلید و درجه ورودی یا خروجی به عنوان مقدار ذخیره می‌شود. سپس با استفاده  از مقادیر هاب و اَثرَیتی که قبلاً محاسبه شده‌اند (از طریق تابع nx.hits) استفاده شده است. برای هر گره در شبکه G، نمرات هاب و authority به ترتیب در hubs_vector و authorities_vector ذخیره می‌شوند. و بردار های ورودی و خروجی به هر نود نیز ایجاد میشوند. با استفاده از متد corrcoef در کتابخانه نامپای میزان همبستگی بین بردار هاب و بردار یال های خروجی و بین بردار authority و بردار یال های ورودی محاسبه شده که مقادر 0.19, 0.29 شده و چون این مقادیر کم هستند نشان دهنده این است که بردار های هاب و authority باید تغییر کنند.
این تغییر می تواند با تنظیم وزن و نرمال سازی بردارها، افزایش تکرار و یا تغییر ساختار گراف باشد.
این قسمت از کد با استفاده از راهنمایی های chatgpt نوشته شده است که پرامپت آن در ادامه آمده است.
پرامپت: https://chatgpt.com/share/673a2523-2518-800d-99a8-a59d7c6acc84
<br><br>
برای قسمت ج من فکر میکنم اگر Hub یک محصول بالا باشد (یعنی بسیاری از کاربران به آن ارجاع داده‌اند یا آن را بازدید کرده‌اند)، احتمالاً این محصول برای کاربران جذاب است و authority بیشتری نیز خواهد داشت (یعنی این محصول توسط کاربران بیشتری خریداری یا بررسی شده است).
authority یک کاربر بالا (یعنی کاربری که بیشترین تعداد محصول را ارزیابی یا خریداری کرده) نشان‌دهنده این است که این کاربر احتمالاً یک hub خوب نیز است، زیرا به احتمال زیاد محصولات بیشتری را بررسی یا خریداری کرده و در نتیجه به دیگر کاربران ارجاع می‌دهد به همین منظور من فکر میکنم اگر همبستگی زیاد باشد، نشان‌دهنده این است که محصولاتی که hub بالایی دارند، بیشتر توسط کاربران تکرار می‌شوند و  کاربران با authority بالا، بیشتر به محصولات مراجعه یا آنها را تکرار می‌کنند. از آنجایی که با محاسبه همبستگی فهمیدیم این مقدار پایین است پس اگر بتوانیم این همبستگی را افزایش دهیم، تاثیر میزان تکرار محصولات و کاربران بر دو بردار hubs , authorities بیشتر خواهد شد.

In [49]:

hubs, authorities = nx.hits(G, max_iter=10)


print("\nTop 10 Hubs:")
top_hubs = sorted(hubs.items(), key=lambda x: x[1], reverse=True)[:10]
for node, score in top_hubs:
    print(f"{node}: {score}")


print("\nTop 10 Authorities:")
top_authorities = sorted(authorities.items(), key=lambda x: x[1], reverse=True)[:10]
for node, score in top_authorities:
    print(f"{node}: {score}")



Top 10 Hubs:
A2PSC7LUNIDEAH: 0.0018764810417924256
A2YB7DLC3FOR7W: 0.0018016671658775444
A2RHV42BTJSVON: 0.0017986964707585947
A3SMV35531YME: 0.0017891676668557856
A3HPCRD9RX351S: 0.0017848743938372943
A44OY8EFDM4IP: 0.0017828863186710713
A3QNQQKJTL76H0: 0.0017712743372021713
A2MNB77YGJ3CN0: 0.0017698110280200336
A3HQAPXAJV7PRX: 0.0017601955882452635
A3MFU0GVZUVH3K: 0.0017586976309857086

Top 10 Authorities:
B001RVFEP2: 0.05083047545976075
B006HYLW32: 0.05083047545976075
B007M83302: 0.050830475459760736
B007M832YY: 0.050830475459760736
B0013NUGDE: 0.05083047545976073
B0026KPDG8: 0.050830475459760716
B001RVFERK: 0.050830475459760716
B000VK8AVK: 0.0508304754597607
B0026KNQSA: 0.0508304754597607
B002IEZJMA: 0.0034720664827639254


In [39]:

in_degrees = dict(G.in_degree())  
out_degrees = dict(G.out_degree()) 
import numpy as np

hubs_vector = np.array([hubs[node] for node in G.nodes()])
authorities_vector = np.array([authorities[node] for node in G.nodes()])


in_degree_vector = np.array([in_degrees[node] if node in in_degrees else 0 for node in G.nodes()])
out_degree_vector = np.array([out_degrees[node] if node in out_degrees else 0 for node in G.nodes()])


correlation_hub_outdegree = np.corrcoef(hubs_vector, out_degree_vector)[0, 1]
correlation_authority_indegree = np.corrcoef(authorities_vector, in_degree_vector)[0, 1]

print(f"Correlation between Hub Scores and Out-degree: {correlation_hub_outdegree:.2f}")
print(f"Correlation between Authority Scores and In-degree: {correlation_authority_indegree:.2f}")
print(hubs_vector)
print(authorities_vector)


Correlation between Hub Scores and Out-degree: 0.19
Correlation between Authority Scores and In-degree: 0.29
[ 9.67849207e-24 -0.00000000e+00  1.52327322e-22 ... -0.00000000e+00
 -2.30574839e-23 -0.00000000e+00]
[-0.00000000e+00  2.67219816e-21 -6.32567914e-20 ...  6.21867831e-19
 -6.67359458e-20 -6.36609149e-21]


د

<div style="direction: rtl; text-align: right;">
در این کد من از متد pagerank کتابخانه networkx استفاده کردم.
ابتدا با استفاده از این متد پیج رنک را بر حسب آلفا برابر ۰.۸۵ با تکرار ۱۰ بار محاسبه کردم.
آلفا میتواند از ۰ تا ۱ باشد ولی معمولا مقدار آن را ۰.۸۵ در نظر میگیرند.
در نهایت، تابع nx.pagerank() نمرات PageRank برای تمامی گره‌ها در گراف G را محاسبه کرده و در دیکشنری pagerank_scores ذخیره می‌کند. هر نود در گراف به یک نمره PageRank مربوط می‌شود که نشان‌دهنده اهمیت آن نود است. در پایان ابتدا، لیستی از گره‌ها و نمرات PageRank آن‌ها که در df['ProductId'] (که حاوی شناسه‌های محصولات است) موجود هستند، استخراج می‌شود.
سپس این لیست بر اساس نمرات PageRank به ترتیب نزولی مرتب می‌شود.
با استفاده از [:10]، ۱۰  محصول با بیشترین نمره PageRank انتخاب می‌شود و این ۱۰ محصول چاپ میشوند.

In [23]:

pagerank_scores = nx.pagerank(G, alpha=0.85, max_iter=10)


print("\nTop 10 Products by PageRank Scores:")
top_products = sorted(
    [(node, score) for node, score in pagerank_scores.items() if node in df['ProductId'].values],
    key=lambda x: x[1],
    reverse=True
)[:10]

for product, score in top_products:
    print(f"Product ID: {product}, PageRank Score: {score}")

with open('top_pagerank_products.txt', 'w') as pr_file:
    for product, score in top_products:
        pr_file.write(f"Product ID: {product}, PageRank Score: {score}\n")



Top 10 Products by PageRank Scores:
Product ID: B007JFMH8M, PageRank Score: 0.0010292803397553628
Product ID: B003B3OOPA, PageRank Score: 0.0006198100436968366
Product ID: B001EO5Q64, PageRank Score: 0.0005186544419068154
Product ID: B000NMJWZO, PageRank Score: 0.0004824322954386272
Product ID: B006H34CUS, PageRank Score: 0.0004134156896336738
Product ID: B000EVOSE4, PageRank Score: 0.00030257531088067707
Product ID: B006Q820X0, PageRank Score: 0.0003021056302219877
Product ID: B004CLCEDE, PageRank Score: 0.00029757921681993485
Product ID: B000KV61FC, PageRank Score: 0.0002890768130073003
Product ID: B000KV7ZGQ, PageRank Score: 0.0002890768130073003


ه

<div style="direction: rtl; text-align: right;">
برای این سوال من فکر کردم اول کاربرهایی که نظراتی بیش از ۱۰ نظر دارند را جدا کنیم و از بین آن ها به صورت رندم یکی را انتخاب کنیم به همین منظور ابتدا نظرات کاربران را شمردم و در user_review_counts ذخیره کردم و از بین آن ها کاربر هایی را که بیش از ۱۰ نظر داشتند را جدا کردم و سپس با استفاده از کتابخانه random یک کاربر را به صورت رندم انتخاب کردم.
سپس یک دیکشنری‌ به نام personalization ایجاد کردم که برای تمامی نودهای گراف (G) مقدار اولیه ۰ را به‌عنوان نمره شخصی‌سازی قرار می‌دهد.
در مرحله بعد برای هر محصولی که توسط کاربر انتخاب‌شده ارزیابی شده، مقدار آن محصول در دیکشنری personalization به ۱ تغییر داده می‌شود. و سپس  الگوریتم Topic Sensitive PageRank با استفاده از پارامتر personalization (که به محصولات ارزیابی‌شده توسط کاربر اختصاص داده شده است) اجرا می‌شود. این باعث می‌شود که در این الگوریتم، محصولات مرتبط با کاربر (بر اساس ارزیابی‌ها) اولویت بالاتری داشته باشند.
پس از محاسبه نمرات page rank، ده محصول برتر انتخاب می‌شوند. این محصولات باید در دیتافریم df موجود باشند تا در فهرست نهایی قرار گیرند. در مرحله آخر این ۱۰ محصول چاپ شده و سپس نمراتی که کاربر به این ۱۰ محصول داده نیز چاپ میشود. نمراتی که برای هر محصول نمایش داده شده نشان میدهد این الگوریتم تا حد زیادی محصولات با نمرات بالا را در رتبه های بالاتری چاپ کرده، بجز ۲ محصول که امتیاز ۲ دارند، امتیاز بقیه محصولات بالای ۳ بوده و غالب این محصولات هم امتیاز برابر ۵ دارند.


In [27]:

import random

user_review_counts = df['UserId'].value_counts()


users_with_many_reviews = user_review_counts[user_review_counts > 10].index


selected_user = random.choice(users_with_many_reviews)
print(f"Randomly Selected User ID: {selected_user}")

personalization = {node: 0 for node in G.nodes()}
for _, row in df[df['UserId'] == selected_user].iterrows():
    product = row['ProductId']
    if product in G.nodes():
        personalization[product] = 1


topic_sensitive_pagerank = nx.pagerank(G, alpha=0.85, personalization=personalization)


top_products_user = sorted(
    [(node, score) for node, score in topic_sensitive_pagerank.items() if node in df['ProductId'].values],
    key=lambda x: x[1],
    reverse=True
)[:10]

print("\nTop 10 Products for Selected User by Topic-sensitive PageRank:")
for product, score in top_products_user:
    print(f"Product ID: {product}, PageRank Score: {score}")

print("\nUser's Ratings for Top 10 Products:")
user_ratings = df[(df['UserId'] == selected_user) & (df['ProductId'].isin([p[0] for p in top_products_user]))]

for _, row in user_ratings.iterrows():
    print(f"Product ID: {row['ProductId']}, User Rating: {row['Score']}")


Randomly Selected User ID: A1BA72K4VK1UN1

Top 10 Products for Selected User by Topic-sensitive PageRank:
Product ID: B001LG945O, PageRank Score: 0.027027027027028232
Product ID: B001BDDTB2, PageRank Score: 0.027027027027028232
Product ID: B007JFMH96, PageRank Score: 0.027027027027028232
Product ID: B004X3VRLG, PageRank Score: 0.027027027027028232
Product ID: B004JRKEH4, PageRank Score: 0.027027027027028232
Product ID: B005CUU25G, PageRank Score: 0.027027027027028232
Product ID: B008O3G25W, PageRank Score: 0.027027027027028232
Product ID: B002LANN56, PageRank Score: 0.027027027027028232
Product ID: B001BCVY9W, PageRank Score: 0.027027027027028232
Product ID: B0090X8IPM, PageRank Score: 0.027027027027028232

User's Ratings for Top 10 Products:
Product ID: B001LG945O, User Rating: 4
Product ID: B001BDDTB2, User Rating: 5
Product ID: B001BDDTB2, User Rating: 5
Product ID: B007JFMH96, User Rating: 3
Product ID: B004X3VRLG, User Rating: 5
Product ID: B004JRKEH4, User Rating: 5
Product ID: B

# **نکات مهم**

<div dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
    <p><strong>مهلت تحویل بدون جریمه:</strong> ۱۴ آبان ۱۴۰۳</p>
    <p><strong>مهلت تحویل با تاخیر (با جریمه):</strong> ۲۱ آبان ۱۴۰۳</p>
</div>
<h4 dir="rtl" style="font-family: Vazir; width: 85%;">فایل ارسالی شما باید با فرمت زیر نامگذاری شود: <code>IIR_CA3_STUDENTID.ipynb</code></h4>

<h4 dir="rtl" style="font-family: Vazir; width: 85%;">نحوه انجام این تمرین:</h4>
<ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;"> <li>برخی سوالات نیاز به نوشتن کد پایتون و محاسبه نتایج دارند کدها بایستی طبق فایل تمرین به طور کامل نوشته شوند.</li> <li> کدها و تفسیرهای هربخش را به طور مشخص در همین نوت‌بوک بنویسید. سعی کنید هربخش به طور مشخصی جداشده باشد و ساختار نوت‌بوک خوانا باشد.</li>  </ul>

<h4 dir="rtl" style="font-family: Vazir; width: 85%;">صداقت علمی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;"> <li>ما نوت‌بوک‌های تعداد مشخصی از دانشجویان که به صورت تصادفی انتخاب می‌شوند، بررسی خواهیم کرد. این بررسی‌ها اطمینان حاصل می‌کنند که کدی که نوشتید واقعاً پاسخ‌های موجود در نوت‌بوک شما را تولید می‌کند. اگر پاسخ‌های صحیح را در نوت‌بوک خود بدون کدی که واقعاً آن پاسخ‌ها را تولید کند تحویل دهید، این یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> <li>ما همچنین بررسی‌های خودکاری را برای تشخیص سرقت علمی در نوت‌بوک‌های کولب انجام خواهیم داد. کپی کردن کد از دیگران نیز یک مورد جدی از عدم صداقت علمی محسوب می‌شود.</li> </ul>

<h4 dir="rtl" style="font-family: Vazir; width: 85%;">توضیحات تکمیلی:</h4> <ul dir="rtl" style="font-family: Vazir; width: 85%; font-size: 16px;">
<li>
خوانایی و دقت بررسی‌ها در گزارش نهایی از اهمیت ویژه‌ای برخوردار است. به تمرین‌هایی که به صورت کاغذی تحویل داده شوند یا به صورت عکس در سایت بارگذاری شوند، ترتیب اثری داده نخواهد شد.</li>
<li>
 همه‌ی کدهای پیوست گزارش بایستی قابلیت اجرای مجدد داشته باشند. در صورتی که برای اجرا مجدد آن‌ها نیاز به تنظیمات خاصی می‌باشد، بایستی تنظیمات مورد نیاز را نیز در گزارش خود ذکر کنید.  دقت کنید که  تمامی کدها باید توسط شما اجرا شده باشند و نتایج اجرا در فایل کدهای ارسالی مشخص باشد. به کدهایی که نتایج اجرای آن‌ها در فایل ارسالی مشخص نباشد نمره‌ای تعلق نمی‌گیرد.
</li>
<li>
تمرین تا یک هفته بعد از مهلت تعیین شده با تاخیر تحویل گرفته می‌شود. دقت کنید که شما جمعاً برای تمام تکالیف، ۱۴ روز زمان تحویل بدون جریمه دارید که تنها از ۷ روز آن برای هر تمرین می‌توانید استفاده کنید. در صورتی که این ۱۴ روز به اتمام رسیده باشد، به ازای هر روز تاخیر ده درصد جریمه می‌شود.
</li>
<li>توجه کنید این تمرین باید به صورت تک‌نفره انجام شود و پاسخ‌های ارائه شده باید نتیجه فعالیت فرد نویسنده باشد (همفکری و به اتفاق هم نوشتن تمرین نیز ممنوع است). در صورت مشاهده
 تشابه به همه افراد مشارکت‌کننده، نمره تمرین صفر و به استاد گزارش می‌گردد.
 </li>
 <li>برای مطالعه بیشتر درباره‌ی فرمت مارک‌دون می‌توانید از <a href="https://github.com/tajaddini/Persian-Markdown/blob/master/learn-MD.md">این لینک</a> مطالعه کنید.
 </li>

 </ul>
 </div>